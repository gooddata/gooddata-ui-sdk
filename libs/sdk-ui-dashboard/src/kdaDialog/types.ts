// (C) 2025 GoodData Corporation

import { IAttributeDescriptorBody, IDashboardAttributeFilter, IMeasure, ObjRef } from "@gooddata/sdk-model";

/**
 * @internal
 */
export interface IKdaDialogProps {
    /**
     * Locale for the dialog
     */
    locale?: string;
    /**
     * CSS class name for the dialog
     */
    className?: string;

    /**
     * Whether to show the close button
     */
    showCloseButton?: boolean;

    /**
     * Custom close button handler
     */
    onClose?: () => void;
}

/**
 * @internal
 */
export interface IKdaDataPoint {
    /**
     * Date string in ISO format related to granularity
     */
    date: string;
    /**
     * Valuer of metric for given range
     */
    value: number | undefined;
    /**
     * Format of date attribute
     */
    format?: IAttributeDescriptorBody["format"];
}

/**
 * @internal
 */
export interface IKdaDefinition {
    /**
     * Metric to analyze
     */
    metric: IMeasure;
    /**
     * Metrics related to main metric
     */
    metrics?: IMeasure[];
    /**
     * Filters to apply
     */
    filters?: IDashboardAttributeFilter[];
    /**
     * Date attribute
     */
    dateAttribute: ObjRef;
    /**
     * Type of period
     */
    type: KdaPeriodType;
    /**
     * Ranges
     */
    range: [IKdaDataPoint, IKdaDataPoint];
}

/**
 * @internal
 */
export type KdaPeriodType = "same_period_previous_year" | "previous_period";
