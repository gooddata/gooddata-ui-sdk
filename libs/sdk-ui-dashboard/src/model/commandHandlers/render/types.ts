// (C) 2025 GoodData Corporation

/**
 * @beta
 */
export interface RenderingWorkerConfiguration {
    /**
     * Maximum time limit for rendering the dashboard.
     * Somehow in sync with limits of exporter
     * https://github.com/gooddata/gdc-exporters-microservices/blob/master/microservices/visual-exporter-service/src/main/kotlin/com/gooddata/exporters/visual/config/TabSessionConfig.kt
     *
     * Default: 20*60000 (20min).
     * @privateRemarks
     * If changing this value update it also in documentation of {@link requestAsyncRender} command creator and {@link useDashboardAsyncRender} hook.
     */
    maxTimeout: number;

    /**
     * Maximum time limit for the first asynchronous rendering request.
     * If no asynchronous rendering request is fired in this time limit, the dashboard will announce that it is rendered.
     *
     * Default: 5000 (5s).
     */
    asyncRenderRequestedTimeout: number;

    /**
     * Maximum time limit to re-request asynchronous rendering of the component once it's resolved.
     *
     * Default: 2000 (2s).
     */
    asyncRenderResolvedTimeout: number;

    /**
     * Generator of correlation ids
     *
     * Default: uuid4
     */
    correlationIdGenerator: () => string;

    /**
     * Wait for given number of async requests.
     * If provided, we'll not wait full asyncRenderRequestedTimeout and
     * terminate the waiting if we reach the given number
     *
     */
    asyncRenderExpectedCount?: number;

    /**
     * Indicates whether rendering worker is triggered inside export mode
     *
     * Default: false
     */
    isExport?: boolean;
}
