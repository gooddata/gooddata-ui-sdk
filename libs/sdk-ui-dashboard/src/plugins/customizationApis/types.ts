// (C) 2021-2025 GoodData Corporation

export type CustomizerMutationsContext = {
    insight: CustomizerMutationsType[];
    kpi: CustomizerMutationsType[];
    attributeFilter: CustomizerMutationsType[];
    dashboardContent: CustomizerMutationsType[];
    topBar: CustomizerMutationsType[];
    filterBar: CustomizerMutationsType[];
    layout: CustomizerMutationsType[];
    richText: CustomizerMutationsType[];
    title: CustomizerMutationsType[];
    loading: CustomizerMutationsType[];
    visualizationSwitcher: CustomizerMutationsType[];
    visualizationSwitcherToolbar: CustomizerMutationsType[];
    layouts: Record<string, CustomizerMutationsType>;
};

export type CustomizerMutationsType = "tag" | "provider" | "body" | "decorator" | "inserted";

export function createCustomizerMutationsContext(): CustomizerMutationsContext {
    return {
        insight: [],
        filterBar: [],
        topBar: [],
        layout: [],
        richText: [],
        title: [],
        kpi: [],
        loading: [],
        attributeFilter: [],
        visualizationSwitcherToolbar: [],
        visualizationSwitcher: [],
        dashboardContent: [],
        layouts: {},
    };
}
