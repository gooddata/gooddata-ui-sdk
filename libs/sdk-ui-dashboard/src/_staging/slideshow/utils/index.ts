// (C) 2022-2025 GoodData Corporation
import {
    IDashboardLayout,
    isDashboardLayout,
    isInsightWidget,
    isKpiWidget,
    isRichTextWidget,
    isVisualizationSwitcherWidget,
    IWidget,
} from "@gooddata/sdk-model";

/**
 * This function checks if the layout widget contains a visualization switcher somewhere in its structure.
 *
 * @param widget - widget to check
 */
export function containsVisualizationSwitcher<TWidget>(widget: TWidget) {
    if (!isDashboardLayout(widget)) {
        return false;
    }
    for (const section of widget.sections) {
        for (const item of section.items) {
            if (isDashboardLayout(item.widget)) {
                return containsVisualizationSwitcher(item.widget);
            } else if (isVisualizationSwitcherWidget(item.widget)) {
                return true;
            }
        }
    }
    return false;
}

function getAllWidgetsFromLayout<TWidget>(layout: IDashboardLayout<TWidget>): IWidget[] {
    return layout.sections.flatMap((section) =>
        section.items.flatMap((item) => {
            const widget = item.widget;
            if (!widget) return [];

            if (isKpiWidget(widget) || isInsightWidget(widget) || isRichTextWidget(widget)) {
                return [widget];
            } else if (isVisualizationSwitcherWidget(widget)) {
                return [widget, ...widget.visualizations];
            } else if (isDashboardLayout(widget)) {
                return [...getAllWidgetsFromLayout(widget)];
            } else {
                return [];
            }
        }),
    );
}

/**
 * This function is used to find a focused widget in the layout structure.
 *
 * @param layout - dashboard layout
 * @param widgetId - widget identifier to find
 * @returns
 */
export const findFocusedWidget = <TWidget>(
    layout: IDashboardLayout<TWidget>,
    widgetId: string | undefined,
): TWidget | undefined => {
    if (!widgetId) {
        return undefined;
    }

    const allWidgets = getAllWidgetsFromLayout(layout);
    return allWidgets.find((widget) => widget.identifier === widgetId) as TWidget;
};
