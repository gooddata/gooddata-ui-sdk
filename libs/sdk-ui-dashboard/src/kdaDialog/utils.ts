// (C) 2025 GoodData Corporation

import { invariant } from "ts-invariant";

import { ClientFormatterFacade, ISeparators } from "@gooddata/number-formatter";
import {
    ICatalogAttribute,
    IDashboardAttributeFilter,
    ObjRef,
    areObjRefsEqual,
    isAttributeElementsByRef,
    isAttributeElementsByValue,
    objRefToString,
} from "@gooddata/sdk-model";
import { DateFilterHelpers } from "@gooddata/sdk-ui-filters";

import { KdaDateOptions } from "./internalTypes.js";
import { IKdaDefinition } from "./types.js";
import { DEFAULT_MEASURE_FORMAT } from "../presentation/alerting/DefaultAlertingDialog/constants.js";

//Format value

export function formatValue(definition: IKdaDefinition, value: number, separators?: ISeparators) {
    const val = ClientFormatterFacade.formatValue(
        value,
        definition.metric.measure.format ?? DEFAULT_MEASURE_FORMAT,
        separators,
    );

    return val.formattedValue;
}

//Format title

export function formatTitle(option: KdaDateOptions, splitter: string): string {
    const [from, to] = option.range ?? [null, null];

    // Not defined
    if (!from || !to) {
        return " - ";
    }

    const pattern = from.format?.pattern ?? "yyyy-MM-dd";

    return DateFilterHelpers.formatAbsoluteDateRange(
        new Date(from.date),
        new Date(to.date),
        pattern,
        splitter,
    );
}

//Attribute filters

export function getOnlyRelevantFilters(all: IDashboardAttributeFilter[], relevantAttrs: ICatalogAttribute[]) {
    return all.filter((f) => {
        return relevantAttrs.some(
            (r) =>
                areObjRefsEqual(f.attributeFilter.displayForm, r.attribute) ||
                r.displayForms.some((df) => areObjRefsEqual(f.attributeFilter.displayForm, df.ref)),
        );
    });
}

export function createNewAttributeFilter(
    attr: ICatalogAttribute,
    displayForm: ObjRef,
    id: string,
    value?: string,
) {
    const displayFormOf = attr.displayForms.find((df) => areObjRefsEqual(df.ref, displayForm));
    invariant(displayFormOf);

    const dashboardFilter: IDashboardAttributeFilter = {
        attributeFilter: {
            displayForm: displayFormOf.ref,
            negativeSelection: !value,
            attributeElements: {
                values: value ? [value] : [],
            },
            localIdentifier: `${objRefToString(displayFormOf.ref)}_${id}`,
            title: displayFormOf.title,
            selectionMode: "multi",
        },
    };

    return dashboardFilter;
}

export function updateExistingAttributeFilter(f: IDashboardAttributeFilter, val: string) {
    return {
        ...f,
        attributeFilter: {
            ...f.attributeFilter,
            negativeSelection: false,
            attributeElements: {
                ...(isAttributeElementsByRef(f.attributeFilter.attributeElements)
                    ? {
                          uris: [val],
                      }
                    : {}),
                ...(isAttributeElementsByValue(f.attributeFilter.attributeElements)
                    ? {
                          values: [val],
                      }
                    : {}),
            },
        },
    } as IDashboardAttributeFilter;
}
