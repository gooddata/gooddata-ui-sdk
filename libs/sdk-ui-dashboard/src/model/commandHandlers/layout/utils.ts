// (C) 2022 GoodData Corporation
import { IDashboardLayoutSectionHeader } from "@gooddata/sdk-model";

/**
 * @internal
 */
export function sanitizeHeader<T extends IDashboardLayoutSectionHeader | undefined>(header: T): T {
    // get rid of empty values in header altogether
    return (header && {
        ...(header.description && { description: header.description }),
        ...(header.title && { title: header.title }),
    }) as T;
}
