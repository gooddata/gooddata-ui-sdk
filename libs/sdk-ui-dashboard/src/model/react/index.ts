// (C) 2021 GoodData Corporation
export {
    DashboardStoreProvider,
    useDashboardDispatch,
    useDashboardSelector,
    ReactDashboardContext,
} from "./DashboardStoreProvider";
export { IDashboardEventsContext, useDashboardEventsContext } from "./DashboardEventsContext";
export { useDashboardCommandProcessing, CommandProcessingStatus } from "./useDashboardCommandProcessing";
export { useDashboardEventDispatch } from "./useDashboardEventDispatch";
export { useDashboardQueryProcessing, QueryProcessingStatus } from "./useDashboardQueryProcessing";
export { useDashboardUserInteraction } from "./useDashboardUserInteraction";
export { useDashboardAsyncRender, UseDashboardAsyncRender } from "./useDashboardAsyncRender";
export { IDashboardStoreProviderProps } from "./types";
export { useDispatchDashboardCommand } from "./useDispatchDashboardCommand";
export { useWidgetExecutionsHandler } from "./useWidgetExecutionsHandler";
