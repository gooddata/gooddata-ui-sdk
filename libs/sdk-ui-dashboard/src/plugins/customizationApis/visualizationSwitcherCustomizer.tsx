// (C) 2021-2025 GoodData Corporation
import { InvariantError } from "ts-invariant";

import { IVisualizationSwitcherCustomizer } from "../customizer.js";
import {
    DefaultDashboardVisualizationSwitcher,
    DefaultVisualizationSwitcherToolbar,
    OptionalVisualizationSwitcherComponentProvider,
    OptionalVisualizationSwitcherToolbarComponentProvider,
    VisualizationSwitcherComponentProvider,
    VisualizationSwitcherToolbarComponentProvider,
} from "../../presentation/index.js";

import { IDashboardCustomizationLogger } from "./customizationLogging.js";
import { CustomizerMutationsContext } from "./types.js";
import union from "lodash/union.js";

const DefaultVisualizationSwitcherRendererProvider: VisualizationSwitcherComponentProvider = () => {
    return DefaultDashboardVisualizationSwitcher;
};

const DefaultVisualizationSwitcherToolbarRendererProvider: VisualizationSwitcherToolbarComponentProvider =
    () => {
        return DefaultVisualizationSwitcherToolbar;
    };

interface IVisualizationSwitcherCustomizerState {
    addCustomProvider(provider: OptionalVisualizationSwitcherComponentProvider): void;
    getRootProvider(): VisualizationSwitcherComponentProvider;
    switchRootProvider(provider: VisualizationSwitcherComponentProvider): void;
}

interface IVisualizationSwitcherToolbarCustomizerState {
    addCustomProvider(provider: OptionalVisualizationSwitcherToolbarComponentProvider): void;
    getRootProvider(): VisualizationSwitcherToolbarComponentProvider;
    switchRootProvider(provider: VisualizationSwitcherToolbarComponentProvider): void;
}

class DefaultVisualizationSwitcherCustomizerState implements IVisualizationSwitcherCustomizerState {
    /*
     * Chain of 'core' providers. Core providers are evaluated from last to first. As soon as some provider
     * returns a Component, then that component will be used for rendering. If provider returns undefined,
     * the evaluation continues to next provider in the chain.
     *
     * Note: the chain is 'primed' with a provider that always returns the default DateFilter renderer. This is
     * essential to allow the decorations to work - purely because decorations can only work if there is
     * something to decorate. See constructor.
     */
    private readonly coreProviderChain: VisualizationSwitcherComponentProvider[];

    /*
     * Core provider encapsulates resolution using the chain of core providers.
     */
    private readonly coreProvider: VisualizationSwitcherComponentProvider = (topBar) => {
        const providerStack = [...this.coreProviderChain].reverse();

        for (const provider of providerStack) {
            const Component = provider(topBar);

            if (Component) {
                return Component;
            }
        }

        // if this happens then the provider chain got messed up. by default the chain contains the default
        // provider which never returns undefined
        throw new InvariantError();
    };

    /*
     * Root provider is THE provider that should be used in the dashboard extension properties. The
     * provider function included here will reflect the setup where there may be N registered decorators
     * sitting on top of a chain of core providers.
     *
     * In the initial state the root provider IS the core provider - meaning no decorations. As the
     * decorators get registered, the rootProvider changes.
     */
    private rootProvider: VisualizationSwitcherComponentProvider = this.coreProvider;

    constructor(defaultProvider: VisualizationSwitcherComponentProvider) {
        this.coreProviderChain = [defaultProvider];
    }

    addCustomProvider(provider: VisualizationSwitcherComponentProvider): void {
        this.coreProviderChain.push(provider);
    }

    getRootProvider(): VisualizationSwitcherComponentProvider {
        return this.rootProvider;
    }

    switchRootProvider(provider: VisualizationSwitcherComponentProvider): void {
        this.rootProvider = provider;
    }
}

/**
 * Sealed customizer state will not allow to perform any modifications of the state. Sealing the state
 * is essential to make the customization more defensive: during the registration, the plugin can hang
 * onto the dashboard customizer - stash it somewhere. And then after registration try to use the customizer
 * and try to do additional 'ad-hoc' customizations.
 *
 * Such a thing is not supported by the lifecycle and this sealing is in place to prevent plugins going into
 * that dangerous territory.
 */
class SealedVisualizationSwitcherCustomizerState implements IVisualizationSwitcherCustomizerState {
    constructor(
        private readonly logger: IDashboardCustomizationLogger,
        private readonly state: IVisualizationSwitcherCustomizerState,
    ) {}

    public addCustomProvider = (_provider: VisualizationSwitcherComponentProvider): void => {
        // eslint-disable-next-line no-console
        this.logger.warn(
            `Attempting to customize VisualizationSwitcher rendering outside of plugin registration. Ignoring.`,
        );
    };

    // eslint-disable-next-line sonarjs/no-identical-functions
    public switchRootProvider = (_provider: VisualizationSwitcherComponentProvider): void => {
        // eslint-disable-next-line no-console
        this.logger.warn(
            `Attempting to customize VisualizationSwitcher rendering outside of plugin registration. Ignoring.`,
        );
    };

    public getRootProvider = (): VisualizationSwitcherComponentProvider => {
        return this.state.getRootProvider();
    };
}

class DefaultVisualizationSwitcherToolbarCustomizerState
    implements IVisualizationSwitcherToolbarCustomizerState
{
    /*
     * Chain of 'core' providers. Core providers are evaluated from last to first. As soon as some provider
     * returns a Component, then that component will be used for rendering. If provider returns undefined,
     * the evaluation continues to next provider in the chain.
     *
     * Note: the chain is 'primed' with a provider that always returns the default DateFilter renderer. This is
     * essential to allow the decorations to work - purely because decorations can only work if there is
     * something to decorate. See constructor.
     */
    private readonly coreProviderChain: VisualizationSwitcherToolbarComponentProvider[];

    /*
     * Core provider encapsulates resolution using the chain of core providers.
     */
    private readonly coreProvider: VisualizationSwitcherToolbarComponentProvider = (topBar) => {
        const providerStack = [...this.coreProviderChain].reverse();

        for (const provider of providerStack) {
            const Component = provider(topBar);

            if (Component) {
                return Component;
            }
        }

        // if this happens then the provider chain got messed up. by default the chain contains the default
        // provider which never returns undefined
        throw new InvariantError();
    };

    /*
     * Root provider is THE provider that should be used in the dashboard extension properties. The
     * provider function included here will reflect the setup where there may be N registered decorators
     * sitting on top of a chain of core providers.
     *
     * In the initial state the root provider IS the core provider - meaning no decorations. As the
     * decorators get registered, the rootProvider changes.
     */
    private rootProvider: VisualizationSwitcherToolbarComponentProvider = this.coreProvider;

    constructor(defaultProvider: VisualizationSwitcherToolbarComponentProvider) {
        this.coreProviderChain = [defaultProvider];
    }

    addCustomProvider(provider: VisualizationSwitcherToolbarComponentProvider): void {
        this.coreProviderChain.push(provider);
    }

    getRootProvider(): VisualizationSwitcherToolbarComponentProvider {
        return this.rootProvider;
    }

    switchRootProvider(provider: VisualizationSwitcherToolbarComponentProvider): void {
        this.rootProvider = provider;
    }
}

/**
 * Sealed customizer state will not allow to perform any modifications of the state. Sealing the state
 * is essential to make the customization more defensive: during the registration, the plugin can hang
 * onto the dashboard customizer - stash it somewhere. And then after registration try to use the customizer
 * and try to do additional 'ad-hoc' customizations.
 *
 * Such a thing is not supported by the lifecycle and this sealing is in place to prevent plugins going into
 * that dangerous territory.
 */
class SealedVisualizationSwitcherToolbarCustomizerState
    implements IVisualizationSwitcherToolbarCustomizerState
{
    constructor(
        private readonly logger: IDashboardCustomizationLogger,
        private readonly state: IVisualizationSwitcherToolbarCustomizerState,
    ) {}

    public addCustomProvider = (_provider: VisualizationSwitcherToolbarComponentProvider): void => {
        // eslint-disable-next-line no-console
        this.logger.warn(
            `Attempting to customize VisualizationSwitcher rendering outside of plugin registration. Ignoring.`,
        );
    };

    // eslint-disable-next-line sonarjs/no-identical-functions
    public switchRootProvider = (_provider: VisualizationSwitcherToolbarComponentProvider): void => {
        // eslint-disable-next-line no-console
        this.logger.warn(
            `Attempting to customize VisualizationSwitcher rendering outside of plugin registration. Ignoring.`,
        );
    };

    public getRootProvider = (): VisualizationSwitcherToolbarComponentProvider => {
        return this.state.getRootProvider();
    };
}

/**
 * Default implementation of the VisualizationSwitcher rendering customizer. Notice that the state of the customizations
 * is kept separate from this class.
 *
 * This code is responsible for creation of the providers (if needed) and then updating the state
 * accordingly. The customizer state methods are responsible for doing correct updates of the state itself. This
 * decoupling is in place so that it is possible to seal the state and prevent write operations from some point
 * onward. See {@link SealedVisualizationSwitcherCustomizerState} for more motivation behind this.
 *
 * @internal
 */
export class DefaultVisualizationSwitcherCustomizer implements IVisualizationSwitcherCustomizer {
    private readonly logger: IDashboardCustomizationLogger;
    private readonly mutationContext: CustomizerMutationsContext;
    private switcherState: IVisualizationSwitcherCustomizerState;
    private toolbarState: IVisualizationSwitcherToolbarCustomizerState;

    constructor(
        logger: IDashboardCustomizationLogger,
        mutationContext: CustomizerMutationsContext,
        defaultSwitcherProvider: VisualizationSwitcherComponentProvider = DefaultVisualizationSwitcherRendererProvider,
        defaultToolbarProvider: VisualizationSwitcherToolbarComponentProvider = DefaultVisualizationSwitcherToolbarRendererProvider,
    ) {
        this.logger = logger;
        this.mutationContext = mutationContext;
        this.switcherState = new DefaultVisualizationSwitcherCustomizerState(defaultSwitcherProvider);
        this.toolbarState = new DefaultVisualizationSwitcherToolbarCustomizerState(defaultToolbarProvider);
    }

    public withCustomSwitcherProvider = (
        provider: OptionalVisualizationSwitcherComponentProvider,
    ): IVisualizationSwitcherCustomizer => {
        this.switcherState.addCustomProvider(provider);
        this.mutationContext.visualizationSwitcher = union(this.mutationContext.visualizationSwitcher, [
            "provider",
        ]);

        return this;
    };

    public withCustomSwitcherDecorator(
        providerFactory: (
            next: VisualizationSwitcherComponentProvider,
        ) => OptionalVisualizationSwitcherComponentProvider,
    ): IVisualizationSwitcherCustomizer {
        // snapshot current root provider
        const rootSnapshot = this.switcherState.getRootProvider();
        // call user's factory in order to obtain the actual provider - pass the current root so that user's
        // code can use it to obtain component to decorate
        const decoratorProvider = providerFactory(rootSnapshot);
        // construct new root provider; this will be using user's provider with a fallback to root provider
        // in case user's code does not return anything
        const newRootProvider: VisualizationSwitcherComponentProvider = (props) => {
            const Component = decoratorProvider(props);

            if (Component) {
                return Component;
            }

            return rootSnapshot(props);
        };

        // finally modify the root provider; next time someone registers decorator, it will be on top of
        // this currently registered one
        this.switcherState.switchRootProvider(newRootProvider);
        this.mutationContext.visualizationSwitcher = union(this.mutationContext.visualizationSwitcher, [
            "decorator",
        ]);

        return this;
    }

    public withCustomToolbarProvider = (
        provider: OptionalVisualizationSwitcherToolbarComponentProvider,
    ): IVisualizationSwitcherCustomizer => {
        this.toolbarState.addCustomProvider(provider);
        this.mutationContext.visualizationSwitcherToolbar = union(
            this.mutationContext.visualizationSwitcherToolbar,
            ["provider"],
        );

        return this;
    };

    public withCustomToolbarDecorator(
        providerFactory: (
            next: VisualizationSwitcherToolbarComponentProvider,
        ) => OptionalVisualizationSwitcherToolbarComponentProvider,
    ): IVisualizationSwitcherCustomizer {
        // snapshot current root provider
        const rootSnapshot = this.toolbarState.getRootProvider();
        // call user's factory in order to obtain the actual provider - pass the current root so that user's
        // code can use it to obtain component to decorate
        const decoratorProvider = providerFactory(rootSnapshot);
        // construct new root provider; this will be using user's provider with a fallback to root provider
        // in case user's code does not return anything
        // eslint-disable-next-line sonarjs/no-identical-functions
        const newRootProvider: VisualizationSwitcherToolbarComponentProvider = (props) => {
            const Component = decoratorProvider(props);

            if (Component) {
                return Component;
            }

            return rootSnapshot(props);
        };

        // finally modify the root provider; next time someone registers decorator, it will be on top of
        // this currently registered one
        this.toolbarState.switchRootProvider(newRootProvider);
        this.mutationContext.visualizationSwitcherToolbar = union(
            this.mutationContext.visualizationSwitcherToolbar,
            ["decorator"],
        );

        return this;
    }

    public getVisualizationSwitcherComponentProvider = (): OptionalVisualizationSwitcherComponentProvider => {
        return this.switcherState.getRootProvider();
    };

    public getVisualizationSwitcherToolbarComponentProvider =
        (): OptionalVisualizationSwitcherToolbarComponentProvider => {
            return this.toolbarState.getRootProvider();
        };

    public sealCustomizer = (): void => {
        this.switcherState = new SealedVisualizationSwitcherCustomizerState(this.logger, this.switcherState);
        this.toolbarState = new SealedVisualizationSwitcherToolbarCustomizerState(
            this.logger,
            this.toolbarState,
        );
    };
}
