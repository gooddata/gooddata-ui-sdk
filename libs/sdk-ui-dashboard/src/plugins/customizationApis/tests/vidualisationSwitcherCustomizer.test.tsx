// (C) 2022-2025 GoodData Corporation
import React from "react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { idRef, IVisualizationSwitcherWidget } from "@gooddata/sdk-model";
import { render } from "@testing-library/react";
import { invariant } from "ts-invariant";

import { DefaultVisualizationSwitcherCustomizer } from "../visualizationSwitcherCustomizer";
import { CustomizerMutationsContext, createCustomizerMutationsContext } from "../types";

import { TestingDashboardCustomizationLogger } from "./fixtures/TestingDashboardCustomizationLogger.js";
import { EMPTY_MUTATIONS } from "./utils";

//
//
//

const TestVisualizationSwitcher: IVisualizationSwitcherWidget = {
    type: "visualizationSwitcher",
    title: "Test RichText",
    ref: idRef("testKpi"),
    identifier: "testKpi",
    description: "Test RichText Widget",
    uri: "/uri/testKpi",
    visualizations: [],
    drills: [],
    ignoreDashboardFilters: [],
};

function renderSwitcherToHtml(customizer: DefaultVisualizationSwitcherCustomizer) {
    const VisualisationSwitcherComponent =
        customizer.getVisualizationSwitcherComponentProvider()(TestVisualizationSwitcher);

    // this should not happen; if it does something is seriously hosed in the customizer
    invariant(VisualisationSwitcherComponent);

    const { container } = render(
        <VisualisationSwitcherComponent screen="lg" widget={TestVisualizationSwitcher} />,
    );

    return container.innerHTML;
}
function renderToolbarToHtml(customizer: DefaultVisualizationSwitcherCustomizer) {
    const VisualisationToolbarComponent =
        customizer.getVisualizationSwitcherToolbarComponentProvider()(TestVisualizationSwitcher);

    // this should not happen; if it does something is seriously hosed in the customizer
    invariant(VisualisationToolbarComponent);

    const { container } = render(
        <VisualisationToolbarComponent
            onClose={vi.fn()}
            onSelectedVisualizationChanged={vi.fn()}
            onVisualizationAdded={vi.fn()}
            onVisualizationsChanged={vi.fn()}
            onWidgetDelete={vi.fn()}
            widget={TestVisualizationSwitcher}
        />,
    );

    return container.innerHTML;
}

describe("visualisation switcher customizer", () => {
    let Customizer: DefaultVisualizationSwitcherCustomizer;
    let mockWarn: ReturnType<typeof vi.fn>;
    let mutationContext: CustomizerMutationsContext;

    beforeEach(() => {
        mockWarn = vi.fn();
        mutationContext = createCustomizerMutationsContext();
        Customizer = new DefaultVisualizationSwitcherCustomizer(
            new TestingDashboardCustomizationLogger({ warn: mockWarn }),
            mutationContext,
            () => {
                function Comp() {
                    return <div>Switcher</div>;
                }
                return Comp;
            },
            () => {
                function Comp() {
                    return <div>Toolbar</div>;
                }
                return Comp;
            },
        );
    });

    describe("visualisation switcher rendering mode", () => {
        it("should render with decorator", () => {
            Customizer.withCustomSwitcherDecorator((_next) => {
                return (_props) => {
                    function Comp() {
                        return <div>Switcher Decorator</div>;
                    }
                    return Comp;
                };
            });

            expect(renderSwitcherToHtml(Customizer)).toMatchSnapshot();
            expect(mutationContext).toEqual({
                ...EMPTY_MUTATIONS,
                visualizationSwitcher: ["decorator"],
            });
        });

        it("should render with provider", () => {
            Customizer.withCustomSwitcherProvider((_props) => {
                function Comp() {
                    return <div>Switcher Provider</div>;
                }
                return Comp;
            });

            expect(renderSwitcherToHtml(Customizer)).toMatchSnapshot();
            expect(mutationContext).toEqual({
                ...EMPTY_MUTATIONS,
                visualizationSwitcher: ["provider"],
            });
        });

        it("should render with default", () => {
            expect(renderSwitcherToHtml(Customizer)).toMatchSnapshot();
            expect(mutationContext).toEqual({
                ...EMPTY_MUTATIONS,
            });
        });
    });

    describe("visualisation switcher toolbar rendering mode", () => {
        it("should render with decorator", () => {
            Customizer.withCustomToolbarDecorator((_next) => {
                return (_props) => {
                    function Comp() {
                        return <div>Toolbar Decorator</div>;
                    }
                    return Comp;
                };
            });

            expect(renderToolbarToHtml(Customizer)).toMatchSnapshot();
            expect(mutationContext).toEqual({
                ...EMPTY_MUTATIONS,
                visualizationSwitcherToolbar: ["decorator"],
            });
        });

        it("should render with provider", () => {
            Customizer.withCustomToolbarProvider((_props) => {
                function Comp() {
                    return <div>Toolbar Provider</div>;
                }
                return Comp;
            });

            expect(renderToolbarToHtml(Customizer)).toMatchSnapshot();
            expect(mutationContext).toEqual({
                ...EMPTY_MUTATIONS,
                visualizationSwitcherToolbar: ["provider"],
            });
        });

        it("should render with default", () => {
            expect(renderToolbarToHtml(Customizer)).toMatchSnapshot();
            expect(mutationContext).toEqual({
                ...EMPTY_MUTATIONS,
            });
        });
    });
});
