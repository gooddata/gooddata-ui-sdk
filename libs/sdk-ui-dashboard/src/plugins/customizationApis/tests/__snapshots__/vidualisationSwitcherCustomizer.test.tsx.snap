// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`visualisation switcher customizer > visualisation switcher rendering mode > should render with decorator 1`] = `"<div>Switcher Decorator</div>"`;

exports[`visualisation switcher customizer > visualisation switcher rendering mode > should render with default 1`] = `"<div>Switcher</div>"`;

exports[`visualisation switcher customizer > visualisation switcher rendering mode > should render with provider 1`] = `"<div>Switcher Provider</div>"`;

exports[`visualisation switcher customizer > visualisation switcher toolbar rendering mode > should render with decorator 1`] = `"<div>Toolbar Decorator</div>"`;

exports[`visualisation switcher customizer > visualisation switcher toolbar rendering mode > should render with default 1`] = `"<div>Toolbar</div>"`;

exports[`visualisation switcher customizer > visualisation switcher toolbar rendering mode > should render with provider 1`] = `"<div>Toolbar Provider</div>"`;
